Sources/TuistLoader/Loaders/CachedManifestLoader.swift

```diff
@@ -0,0 +1,183 @@
+    private func calculateHelpersCache(at path: AbsolutePath) throws -> String? {
+        guard let helpersDirectory = helpersDirectoryLocator.locate(at: path) else {
+            return nil
+        }
+
+        if let cached = helpersCache[helpersDirectory] {
+            return cached
+        }
+
+        let hash = try projectDescriptionHelpersHasher.hash(helpersDirectory: helpersDirectory)
+        helpersCache[helpersDirectory] = hash
+
+        return hash
+    }
+
+    private func cachedPath(for manifestPath: AbsolutePath) -> AbsolutePath {
+        let pathHash = manifestPath.pathString.md5
+        let cacheVersion = CachedManifest.currentVersion.description
+        let fileName = [cacheVersion, pathHash].joined(separator: ".")
+        return cacheDirectory.appending(component: fileName)
+    }
+
+    private func loadCachedManifest<T: Decodable>(at cachedManifestPath: AbsolutePath,
+                                                  manifestHash: Data,
+                                                  helpersHash: String?) -> T? {
+        guard fileHandler.exists(cachedManifestPath) else {
+            return nil
+        }
+
+        guard let data = try? fileHandler.readFile(cachedManifestPath) else {
+            return nil
+        }
+
+        guard let cachedManifest = try? decoder.decode(CachedManifest.self, from: data) else {
+            return nil
+        }
+
+        guard cachedManifest.version == CachedManifest.currentVersion,
+            cachedManifest.helpersHash == helpersHash,
+            cachedManifest.manifestHash == manifestHash else {
+            return nil
+        }
+
+        return try? decoder.decode(T.self, from: cachedManifest.manifest)
+    }
+
+    private func cacheManifest<T: Encodable>(manifest: Manifest,
+                                             loadedManifest: T,
+                                             manifestHash: Data,
+                                             helpersHash: String?,
+                                             to cachedManifestPath: AbsolutePath) throws {
+        let cachedManifest = CachedManifest(manifestHash: manifestHash,
+                                            helpersHash: helpersHash,
+                                            manifest: try encoder.encode(loadedManifest))
+
+        let cachedManifestData = try encoder.encode(cachedManifest)
+        guard let cachedManifestContent = String(data: cachedManifestData, encoding: .utf8) else {
+            throw ManifestLoaderError.manifestCachingFailed(manifest, cachedManifestPath)
+        }
+
+        try fileHandler.touch(cachedManifestPath)
+        try fileHandler.write(
+            cachedManifestContent,
+            path: cachedManifestPath,
+            atomically: true
+        )
+    }
+}
+
+private struct CachedManifest: Codable {
+    static let currentVersion = 1
+    var version: Int = currentVersion
+    var manifestHash: Data
+    var helpersHash: String?
+    var manifest: Data
```

Comment:
I think we should include here the version of Tuist. Since we are not introducing breaking changes in the manifests, it should not make a big difference, but in case we do, we should detect it and rather read the manifest from disk. What about hashing the `ProjectDescription` framework instead of the Tuist version?

Reply:
In the event the manifest parsing fails, we fallback to loading from disk e.g. if a breaking change were to be introduced. 

I think the Tuist version would suffice, the `ProjectDescription` will always be different even for the same Tuist version (each time its built it will have a different hash).